lib/lualink.c

```diff
@@ -537,28 +543,71 @@ void Lua_crowbegin( void )
 {
     printf("init()\n"); // call in C to avoid user seeing in lua
     lua_getglobal(L,"init");
-    if( lua_pcall_protected(L,0,0,0) != LUA_OK ){
-        Caw_send_luachunk( (char*)lua_tostring( L, -1 ) );
-        lua_pop( L, 1 );
+    if( Lua_call_usercode(L,0,0) != LUA_OK ){
+        Caw_send_luachunk("error running init()");
+        lua_pop(L, 1);
     }
 }
 
 
 // Watchdog timer for infinite looped Lua scripts
 volatile int watchdog = WATCHDOG_COUNT;
-static void hook( lua_State* L, lua_Debug* ar )
+static void timeouthook( lua_State* L, lua_Debug* ar )
 {
     if( --watchdog <= 0 ){
         Caw_send_luachunk("CPU timed out.");
-        lua_sethook(L, hook, LUA_MASKLINE, 0); // error until top
-        luaL_error(L,"");
+        lua_sethook(L, timeouthook, LUA_MASKLINE, 0); // error until top
+        luaL_error(L, "user code timeout exceeded");
+    }
+}
+
+// Check for too much recursion
+static void callhook( lua_State* L, lua_Debug* ar )
+{
+    for (int level = 0; lua_getstack(L, level, ar); level++) {
+        if (level > MAX_CALLFRAMES) {
+	    Caw_send_luachunk("Call depth exceeded " TOSTRING(MAX_CALLFRAMES));
+	    luaL_error(L, "recursion limit exceeded");
+	}
     }
```

Comment:
Maybe I'm misunderstanding this logic, but couldn't this just be:
```
if( lua_getstack(L, level, ar) > MAX_CALLFRAMES ){
  // recursion limit exceeded
}
```

Reply:
Yeah this function is probably actually quadratic in the number of call frames! Should just do `if( lua_getstack(L, MAX_CALLFRAMES, ar) ) != 0` -- `lua_getstack` already walks a linked list of function calls and returns 0 if there is no such level.